presto-parquet/src/main/java/io/prestosql/parquet/reader/PrimitiveColumnReader.java

```diff
@@ -78,32 +92,80 @@ protected boolean isValueNull()
-                    return new TimestampMicrosColumnReader(descriptor);
+        requireNonNull(descriptor, "descriptor is null");
+        requireNonNull(prestoType, "prestoType is null");
+
+        Optional<DecimalType> parquetDecimalType = createDecimalType(descriptor);
+        Optional<PrimitiveColumnReader> decimalColumnReader = createDecimalColumnReader(descriptor);
+        verify(parquetDecimalType.isPresent() == decimalColumnReader.isPresent(), "parquetDecimalType and decimalColumnReader should be present at the same time");
+
+        if (prestoType == BooleanType.BOOLEAN && descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.BOOLEAN) {
+            return new BooleanColumnReader(descriptor);
+        }
+
+        if (prestoType == TinyintType.TINYINT
+                || prestoType == SmallintType.SMALLINT
+                || prestoType == IntegerType.INTEGER
+                || prestoType == BigintType.BIGINT) {
+            // TODO verify no truncation occurs
+            if (parquetDecimalType.isPresent() && parquetDecimalType.get().getScale() == 0) {
+                // TODO use IntColumnReader or LongColumnReader when applicable
+                return decimalColumnReader.orElseThrow(() -> new VerifyException("decimalColumnReader not present when parquetDecimalType is present"));
+            }
+
+            if (!parquetDecimalType.isPresent()) {
+                if (descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.INT32) {
+                    return new IntColumnReader(descriptor);
+                }
+                if (descriptor.getPrimitiveType().getPrimitiveTypeName() == PrimitiveTypeName.INT64) {
+                    return new LongColumnReader(descriptor);
                 }
-                return createDecimalColumnReader(descriptor).orElse(new LongColumnReader(descriptor));
-            case INT96:
-                return new TimestampColumnReader(descriptor);
-            case FLOAT:
-                return new FloatColumnReader(descriptor);
-            case DOUBLE:
-                return new DoubleColumnReader(descriptor);
-            case BINARY:
-                return createDecimalColumnReader(descriptor).orElse(new BinaryColumnReader(descriptor));
-            case FIXED_LEN_BYTE_ARRAY:
-                return createDecimalColumnReader(descriptor)
-                        .orElseThrow(() -> new PrestoException(NOT_SUPPORTED, " type FIXED_LEN_BYTE_ARRAY supported as DECIMAL; got " + descriptor.getPrimitiveType().getOriginalType()));
-            default:
-                throw new PrestoException(NOT_SUPPORTED, "Unsupported parquet type: " + descriptor.getPrimitiveType().getPrimitiveTypeName());
+            }
+        }
+
+        if (prestoType instanceof DecimalType) {
+            if (parquetDecimalType.isPresent() && ((DecimalType) prestoType).getScale() == parquetDecimalType.get().getScale()) {
```

Comment:
nit: I think the previous message the explicitly mentioned different scales was nicer

Reply:
scales are being addressed by @wendigo in a separate PR, so let's skip this one